graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -528,38 +530,73 @@ def _rename_or_suppress_or_ignore_name_and_add_to_record(
             - SchemaNameConflictError if the newly renamed node causes name conflicts with
               existing types, scalars, or builtin types
         """
-        name_string = node.name.value
+        type_name = node.name.value
 
-        if name_string == self.query_type or name_string in self.scalar_types:
+        if (
+            type_name == self.query_type
+            or type_name in self.custom_scalar_names
+            or type_name in builtin_scalar_type_names
+        ):
             return IDLE
 
-        new_name_string = self.renamings.get(name_string, name_string)  # Default use original
-        if new_name_string is None:
+        desired_type_name = self.renamings.get(type_name, type_name)  # Default use original
+        if desired_type_name is None:
             # Suppress the type
             return REMOVE
-        check_type_name_is_valid(new_name_string)
-
+        check_type_name_is_valid(desired_type_name)
+
+        # Renaming conflict arises when two types with different names in the original schema have
+        # the same name in the new schema. There are two ways to produce this conflict:
+        # * when neither type is a custom scalar
+        # * when one type is a custom scalar and the other is not
+        #
+        # If neither type is a custom scalar, then desired_type_name will be in
+        # self.reverse_name_map (because self.reverse_name_map records all non-scalar types in
+        # the schema). The types named self.reverse_name_map[desired_type_name] and type_name in
+        # the old schema would both get mapped to desired_type_name in the new schema, even though
+        # self.reverse_name_map[desired_type_name] != type_name.
+        #
+        # If one type in the conflict is a custom scalar, then the conflict arises from
+        # attempting to rename a non-scalar type (from type_name to desired_type_name) when
+        # there already exists a custom scalar type named desired_type_name. Custom scalar
+        # renaming has not been implemented yet so we know for sure that the scalar's name (in
+        # both the original and new schema) is desired_type_name.
+        #
+        # It's also possible to produce a similar conflict where one type is not a custom scalar but
+        # rather a built-in scalar, e.g. String. That case is handled separately because renaming a
+        # type to a built-in scalar will never be allowed in any schema, whereas the conflicts here
+        # arise from conflicting with the existence of other types in the schema.
```

Comment:
This giant comment is super helpful for understanding the code! Thanks for adding 😄